### PR TITLE
Add Pyodide support and CI jobs for Zarr

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
## Description

This PR adds a CI job that builds Pyodide wheels for Zarr and runs the entire test suite against it, skipping unit tests entirely where certain modules from the Python standard library are not available or limited in functionality, which at the time are `fcntl`, `threading`, and `multiprocessing` (see also: https://github.com/pyodide/pyodide/issues/237). I notice ne of the [patches](https://github.com/pyodide/pyodide/blob/main/packages/zarr/patches/0001-Fix-zarr-sync.patch) from the Pyodide repository has already been applied for the Zarr source code; it is not necessary to apply it here, and the rest of the imports can either be guarded with if-else clauses or left as-is, since there are not many uses of threads and tests with that functionality are already disabled.

This PR _might_ not add _in-house_  support for Zarr, and it does not intend to cover all ends, because a lot of the [storage classes](https://zarr.readthedocs.io/en/stable/api/storage.html) inherently rely on threading and multiprocessing to function properly, and as someone who is not a frequent user of Zarr, and more of a novice around it, I may not have enough context about this, because, based on #1777 and #1274, these stores are being removed in version 3. It would also be a large, enterprising project owing to the sheer number of changes required to adapt everything to a non-threaded, serial implementation. I think that would be better as a separate PR.

Here are some related conversations where Pyodide (and its related discussions) has come up as a topic: #536, #1018, #1019, #1104, fsspec/filesystem_spec#960, and zarr-developers/community#43

It is to be noted that this PR has been accompanied by a corresponding PR for out-of-tree Pyodide builds for `numcodecs` at zarr-developers/numcodecs#529, where `numcodecs` builds successfully and almost all of the test suite passes (just a few of them are remaining to be fixed). The change was proposed in one of the community meetings for Zarr this year ([meeting notes](https://hackmd.io/tFcffc__QRuo1UdGcGUbsg#2024-03-06)). However, it is not required to build both packages at the same time since [`numcodecs` already exists in-tree](https://github.com/pyodide/pyodide/tree/main/packages/numcodecs) for Pyodide version 0.25.1 (which is the one being implemented here as well at the time of writing), and therefore `numcodecs` from the in-tree distribution (version 0.11.0) is currently being used for running Zarr's test suite.

Zarr does not require any specific build-time configuration owing to the fact that it is a pure Python package and does not implement any compiled components or CPython extension modules. There is native support for building such packages through the build frontend that ships with Pyodide (`pyodide-build`; a variant that monkey-patches `pypa/build`).

TODO:

* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

## Additional context

Here are a few xrefs to some other PRs that have been implemented for packages in the Scientific Python ecosystem:

- NumPy's Pyodide job: https://github.com/numpy/numpy/pull/25894, by courtesy of @rgommers
- The same for PyWavelets (which is an optional dependency for `scikit-image`): https://github.com/PyWavelets/pywt/pull/701 and https://github.com/PyWavelets/pywt/pull/714
- For the `pandas` repository https://github.com/pandas-dev/pandas/pull/57896
- https://github.com/scikit-image/scikit-image/pull/7350
- and so on

The greater, long-term goal towards this change is to implement Sphinx-based interactive documentation via JupyterLite and WASM-powered, in-browser kernels, as referenced in Quansight-Labs/czi-scientific-python-mgmt#19, see also: Quansight-Labs/czi-scientific-python-mgmt#18. A pilot that can be readily tried out is available for the "API reference" pages under the [PyWavelets documentation](https://pywavelets.readthedocs.io/en/latest/ref/wavelets.html). This will be preceded by configuring a place to publish these WebAssembly/Emscripten wheels nightly or on a scheduled cadence for Zarr (which could be a third-party, PyPI-like index such as Anaconda.org) and then integrating Sphinx extensions such as [`jupyterlite-sphinx`](https://github.com/jupyterlite/jupyterlite-sphinx) for hosted documentation.

### CODE REVIEW

File: zarr/tests/test_core.py [hunk 1]
```diff
@@ -28,7 +28,17 @@
...
+
+try:
+    from numcodecs.tests.common import greetings
+except ModuleNotFoundError:
+    greetings = [
+        '¡Hola mundo!', 'Hej Världen!', 'Servus Woid!', 'Hei maailma!',
+        'Xin chào thế giới', 'Njatjeta Botë!', 'Γεια σου κόσμε!',
+        'こんにちは世界', '世界，你好！', 'Helló, világ!', 'Zdravo svete!',
+        'เฮลโลเวิลด์'
+]
```
Comment (agriyakhetarpal) [inline zarr/tests/test_core.py:40]: I am unaware of the reason why `numcodecs.tests.common` cannot import this to the namespace – for this while, I have added this list of strings verbatim here. I will investigate the cause for the error.

File: zarr/tests/test_storage.py [hunk 2]
```diff
@@ -1020,6 +1020,7 @@ def create_store(self, normalize_keys=False, dimension_separator=".", **kwargs):
         )
         return store
 
+    @pytest.mark.xfail(reason="Skip under Pyodide for now, don't know how to handle umask")
```
Comment (agriyakhetarpal) [inline zarr/tests/test_storage.py:1023]: This is one of the tests I would like to receive help on – I am not entirely aware of the differences in the Pyodide (Emscripten MemFS to be specific) and the CPython file systems, and I see that retrieving the 0o777 octal (execute permissions) seems to bring some insights. I have marked this as xfail for now.

**Review by agriyakhetarpal — COMMENTED** [review]
I added a couple of review comments below to ease the code review process for project maintainers and core developers. Besides that, there is [another test case that fails](https://github.com/agriyakhetarpal/zarr-python/actions/runs/9196692454/job/25295364825#step:8:5406) and I have had a hard time debugging it so far. I have managed to get down to two failing tests across the entire test suite by skipping tests where applicable using `@pytest.mark.skipif()` decorators.

Additionally, I suppose that proper updates to the Zarr documentation and a CHANGELOG update would be required to inform people about the limitations of running Zarr in a Pyodide runtime and which features are not supported. I would be happy to make those changes as I am directed to; and edit the docstrings for the relevant sections in the public API, in subsequent commits and review comments as necessary.

File: pyproject.toml [hunk 3]
```diff
@@ -16,6 +16,8 @@ dependencies = [
     'numpy>=1.23',
     'fasteners; sys_platform != "emscripten"',
     'numcodecs>=0.10.0',
+    'numcodecs[msgpack]>=0.10.0; sys_platform == "emscripten"',  # TODO: can be removed
+    'msgpack; sys_platform == "emscripten"',
```
Comment (agriyakhetarpal) [inline pyproject.toml:20]: Both of these lines currently do not work for some reason – I am unable to import `MsgPack` from `numcodecs` even after I added them, which is why I am explicitly installing `msgpack` inside the Pyodide virtual environment for now, which seems to make this work. I will take a look at why these options do not work as I continue to work on this PR.

**Comment by jhamman** [comment]
@agriyakhetarpal -- thanks for this PR. I want to point you to https://github.com/zarr-developers/zarr-python/issues/1777 and see if we can focus your efforts toward the `v3` branch rather than `main`.

**Comment by agriyakhetarpal** [comment]
Thank you for the redirection, @jhamman. I shall change the base branch of the PR, fix any conflicts, and rebase as necessary in a moment.

**Comment by agriyakhetarpal** [comment]
Okay, wayyy too many conflicts! 😄 I think this will take me a while, but should be doable without repeating these changes as a separate PR.

**Comment by agriyakhetarpal** [comment]
I cherry-picked and backed up my commits in a different branch and a stash, I'll recover them from there. I think a separate PR would be better, I'll copy what I documented over to the other one.

**Comment by agriyakhetarpal** [comment]
I have a suggestion which I hope is not unsolicited: please make the `v3` branch the default branch for the repository in the GitHub settings? It would be helpful in the context of new contributors – my Git TUI used the `main` branch to create a new branch.

Please ignore if this has been evaluated before or if there is a nuance to the discussion that I might have missed as a newcomer who has not contributed to Zarr before.

**Comment by agriyakhetarpal** [comment]
I have a suggestion which I hope is not unsolicited: please make the `v3` branch the default branch for the repository in the GitHub settings? It would be helpful in the context of new contributors – my Git TUI used the `main` branch to create a new branch. :)

Please ignore if this has been evaluated before or if there is a nuance to the discussion that I might have missed as a newcomer who has not contributed to Zarr before. Thanks!

**Pull request closed without merging**